### PR TITLE
deepin(library): init at 20.8

### PR DESCRIPTION
**Pull request by wineee**
###### Description of changes
I'm trying to port  [Deepin Desktop Environment](https://www.deepin.org/en/dde/) to NixOS. Resume work of @romildo and (at)worldofpeace on https://github.com/NixOS/nixpkgs/issues/94870

I created a [flake repository](https://github.com/linuxdeepin/dde-nixos) for this work. Porting work nearing completion，I think it's time to request a review. 

It will need 70+ commits to add DDE for NixOS, Here is the first part, add all libraries needed for DDE.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by onny** [comment]
Result of `nixpkgs-review pr 210477` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>deepin.dde-qt-dbus-factory</li>
    <li>deepin.deepin-wayland-protocols</li>
    <li>deepin.disomaster</li>
    <li>deepin.docparser</li>
    <li>deepin.dtkcommon</li>
    <li>deepin.dtkcore</li>
    <li>deepin.dtkgui</li>
    <li>deepin.dtkwidget</li>
    <li>deepin.dwayland</li>
    <li>deepin.gio-qt</li>
    <li>deepin.image-editor</li>
    <li>deepin.qt5integration</li>
    <li>deepin.qt5platform-plugins</li>
    <li>deepin.udisks2-qt5</li>
  </ul>
</details>

**Comment by wineee** [comment]
@ofborg build deepin

File: pkgs/desktops/deepin/default.nix [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    replaceAll = x: y: ''
+      echo Replacing "${x}" to "${y}":
+      for file in $(grep -rl "${x}")
+      do
+        echo -- $file
+        substituteInPlace $file \
+          --replace "${x}" "${y}"
+      done
+    '';
+  };
```
Comment (NickCao) [inline pkgs/desktops/deepin/default.nix:30]: These functions seem to be unused?

File: pkgs/desktops/deepin/library/deepin-wayland-protocols/default.nix [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{ stdenv
+, lib
+, fetchFromGitHub
+, cmake
+, extra-cmake-modules
+}:
+stdenv.mkDerivation rec {
```
Comment (romildo) [inline pkgs/desktops/deepin/library/deepin-wayland-protocols/default.nix:7]:
```suggestion

stdenv.mkDerivation rec {
```

File: pkgs/desktops/deepin/library/dwayland/default.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+, cmake
+, qtbase
+, qtwayland
+, wayland
+, wayland-protocols
+, extra-cmake-modules
+, deepin-wayland-protocols
+, qttools
+}:
+stdenv.mkDerivation rec {
```
Comment (romildo) [inline pkgs/desktops/deepin/library/dwayland/default.nix:13]:
```suggestion

stdenv.mkDerivation rec {
```

File: pkgs/desktops/deepin/library/qt5integration/default.nix [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    qt5platform-plugins
+    mtdev
+    lxqt.libqtxdg
+    xorg.xcbutilrenderutil
+    gtest
+  ];
+
+  installPhase = ''
+    mkdir -p $out/${qtbase.qtPluginPrefix}
+    cp -r bin/plugins/* $out/${qtbase.qtPluginPrefix}/
```
Comment (romildo) [inline pkgs/desktops/deepin/library/qt5integration/default.nix:49]:
```suggestion
    runHook preInstall
    mkdir -p $out/${qtbase.qtPluginPrefix}
    cp -r bin/plugins/* $out/${qtbase.qtPluginPrefix}/
    runHook postInstall
```

File: pkgs/desktops/deepin/library/image-editor/default.nix [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        --replace 'libPath("libffmpegthumbnailer.so")'  'QString("${ffmpegthumbnailer.out}/lib/libffmpegthumbnailer.so")'
+
+    substituteInPlace libimageviewer/CMakeLists.txt --replace '/usr' '$out'
+    substituteInPlace libimagevisualresult/CMakeLists.txt --replace '/usr' '$out'
+  '';
+
+  nativeBuildInputs = [ cmake pkg-config qttools wrapQtAppsHook ];
+
+  buildInputs = [
+    dtk
```
Comment (romildo) [inline pkgs/desktops/deepin/library/image-editor/default.nix:52]: I am not sure that `buildInputs` can contain a nested list. Do you know if this possibility is documented somewhere?

**Review by romildo — CHANGES_REQUESTED** [review]

File: pkgs/desktops/deepin/default.nix [hunk 6]
```diff
@@ -0,0 +1,52 @@
+{ lib, pkgs, libsForQt5 }:
+let
+  functions = with pkgs.lib; rec {
```
Comment (AndersonTorres) [inline pkgs/desktops/deepin/default.nix:3]: Avoid globally scoped `with lib`.

File: pkgs/desktops/deepin/default.nix [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    getPatchFrom' = commonRp:
+      let
+        rpstr = a: b: " --replace \"${a}\" \"${b}\"";
```
Comment (AndersonTorres) [inline pkgs/desktops/deepin/default.nix:6]: Don't use cryptic names. This is not C code.

File: pkgs/desktops/deepin/library/image-editor/default.nix (hunk 5 above)
Comment (wineee) [reply pkgs/desktops/deepin/library/image-editor/default.nix:52]: Use `propagatedBuildInputs` should be more natural，I will delete this list

File: pkgs/desktops/deepin/default.nix (hunk 1 above)
Comment (wineee) [reply pkgs/desktops/deepin/default.nix:30]: There are too much  hardcode path in dde,  library in this pr does need these functions，but some of DDE core components need this function to generate the patch：

https://github.com/linuxdeepin/dde-nixos/blob/main/packages/core/dde-file-manager/default.nix
https://github.com/linuxdeepin/dde-nixos/blob/main/packages/go-package/dde-daemon/default.nix
https://github.com/linuxdeepin/dde-nixos/blob/main/packages/core/dde-session-shell/default.nix
https://github.com/linuxdeepin/dde-nixos/blob/main/packages/core/dde-session-ui/default.nix

File: pkgs/desktops/deepin/library/image-editor/default.nix [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  };
+
+  patches = [
+    (fetchpatch {
+      name = "feat_check_PREFIX_value_before_set";
+      url = "https://github.com/linuxdeepin/image-editor/commit/dae86e848cf53ba0ece879d81e8d5335d61a7473.patch";
+      sha256 = "sha256-lxmR+nIrMWVyhl1jpA17x2yqJ40h5vnpqKKcjd8j9RY=";
+    })
+    (fetchpatch {
+      name = "feat: use FULL install path";
```
Comment (AndersonTorres) [inline pkgs/desktops/deepin/library/image-editor/default.nix:35]: why put whitespaces here?

File: pkgs/desktops/deepin/library/image-editor/default.nix [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    ffmpegthumbnailer
+    pcre
+  ];
+
+  cmakeFlags = [
+    "-DVERSION=${version}"
+  ];
+
+  meta = with lib; {
+    description = "image editor lib for dtk";
```
Comment (AndersonTorres) [inline pkgs/desktops/deepin/library/image-editor/default.nix:65]:
```suggestion
    description = "Image editor lib for dtk";
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/desktops/deepin/library/image-editor/default.nix (hunk 5 above)
Comment (romildo) [reply pkgs/desktops/deepin/library/image-editor/default.nix:52]: The `dtk` build input should also be removed from `pkgs/desktops/deepin/library/image-editor/default.nix`, I suppose.

**Comment by wineee** [comment]
@ofborg build deepin

File: pkgs/desktops/deepin/default.nix (hunk 7 above)
Comment (wineee) [reply pkgs/desktops/deepin/default.nix]: This functions not used yet, I will adjust this function in the future pr
This function has been removed in this pr

File: pkgs/desktops/deepin/default.nix [hunk 10]
```diff
@@ -0,0 +1,21 @@
+{ lib, pkgs, libsForQt5 }:
+let
+  packages = self: with self; {
```
Comment (AndersonTorres) [inline pkgs/desktops/deepin/default.nix:3]: Letś get rid of "`with`'s that  no one knows how it works":

```suggestion
  packages = self:
  let
    inherit (self) callPackage;
  in {
```

File: pkgs/desktops/deepin/library/deepin-wayland-protocols/default.nix [hunk 11]
```diff
@@ -0,0 +1,31 @@
+{ stdenv
+, lib
+, fetchFromGitHub
+, cmake
+, extra-cmake-modules
+}:
+
+stdenv.mkDerivation rec {
+  pname = "deepin-wayland-protocols";
+  version = "1.6.0-deepin.1.2";
```
Comment (AndersonTorres) [inline pkgs/desktops/deepin/library/deepin-wayland-protocols/default.nix:10]: It can't be used in `src.rev`?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/desktops/deepin/library/deepin-wayland-protocols/default.nix (hunk 11 above)
Comment (wineee) [reply pkgs/desktops/deepin/library/deepin-wayland-protocols/default.nix:10]: dde-kwin/deepin-kwin/deepin-wayland-protocols/dwayland use snipe branch（for deepin v23)，This branch has just been synced to github, but the tag is not synced. so use commit like `chore: Update version number for 5.24.3-deepin.1.9` https://github.com/linuxdeepin/dwayland/commit/3e2d16d4278959c2aa76e1e1f4c0cff9f52a109f

upd:  now the tags of other projects have been synchronized except deepin-kwin

**Comment by wineee** [comment]
@ofborg build deepin

**Review by NickCao — APPROVED** [review]
diff LGTM

**Review by AndersonTorres — APPROVED** [review]
LGTM.

File: pkgs/desktops/deepin/library/dtkgui/default.nix [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  buildInputs = [
+    lxqt.libqtxdg
+  ];
+
+  propagatedBuildInputs = [
+    dtkcore
+    librsvg
+    qtimageformats
+    freeimage
+  ];
```
Comment (SuperSandro2000) [inline pkgs/desktops/deepin/library/dtkgui/default.nix:42]: Do we really need to do this?

File: pkgs/desktops/deepin/library/gio-qt/default.nix [hunk 13]
```diff
@@ -0,0 +1,42 @@
+{ stdenv
+, lib
+, fetchFromGitHub
+, cmake
+, pkg-config
+, wrapQtAppsHook
+, glibmm
+, doxygen
+, buildDocs ? false
```
Comment (SuperSandro2000) [inline pkgs/desktops/deepin/library/gio-qt/default.nix:9]: Whats the reason why this is off by default?

File: pkgs/desktops/deepin/library/gio-qt/default.nix [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+  nativeBuildInputs = [ cmake pkg-config wrapQtAppsHook ];
+
+  cmakeFlags = [
+    "-DCMAKE_INSTALL_LIBDIR=lib"
+    "-DPROJECT_VERSION=${version}"
+  ]
+  ++ lib.optional (!buildDocs) [ "-DBUILD_DOCS=OFF" ];
+
+  buildInputs = lib.optional buildDocs doxygen;
```
Comment (SuperSandro2000) [inline pkgs/desktops/deepin/library/gio-qt/default.nix:31]: doxygen goes to nativeBuildInputs usually

File: pkgs/desktops/deepin/library/dtkgui/default.nix (hunk 12 above)
Comment (NickCao) [reply pkgs/desktops/deepin/library/dtkgui/default.nix:42]: `dtkcore` is required in `dtkgui.pc`, not sure about the rest.

Comment (wineee) [reply pkgs/desktops/deepin/library/dtkgui/default.nix:42]: qtimageformats is a runtime dependency，is uesd for parse dci icons (webp format)
image
librsvg is  runtime dependency uesd by DSvgRenderer
freeimage is  runtime dependency uesd by  DImageHandler

File: pkgs/desktops/deepin/library/gio-qt/default.nix (hunk 13 above)
Comment (wineee) [reply pkgs/desktops/deepin/library/gio-qt/default.nix:9]: fix in  https://github.com/NixOS/nixpkgs/pull/212979

File: pkgs/desktops/deepin/library/dtkgui/default.nix (hunk 12 above)
Comment (SuperSandro2000) [reply pkgs/desktops/deepin/library/dtkgui/default.nix:42]: This cannot be added via the qt wrapper or another variable?

File: pkgs/desktops/deepin/library/qt5platform-plugins/default.nix [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+, fetchFromGitHub
+, qmake
+, pkg-config
+, qtbase
+, qtx11extras
+, wrapQtAppsHook
+, mtdev
+, cairo
+, xorg
+, waylandSupport ? false
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/desktops/deepin/library/qt5platform-plugins/default.nix:12]: Maybe we should add a comment explaining this

[the remark this answers is not in the document]
Comment (wineee) [reply pkgs/desktops/deepin/library/qt5platform-plugins/default.nix:12]: Should I remove this option and disable wayland compilation directly？

[the remark this answers is not in the document]
Comment (wineee) [reply pkgs/desktops/deepin/library/qt5platform-plugins/default.nix:12]: By the way, wayland support in deepin may be complete by the end of this year

File: pkgs/desktops/deepin/library/dtkgui/default.nix (hunk 12 above)
Comment (wineee) [reply pkgs/desktops/deepin/library/dtkgui/default.nix:42]: QT_PLUGIN_PATH will be set if qtimageformats in `buildInputs`  and wrapQtAppsHook in `nativeBuildInputs`
if dont use `propagatedBuildInputs`， we have to set qtimageformats wrapper for all dtk application

File: pkgs/desktops/deepin/library/qt5platform-plugins/default.nix (hunk 15 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/desktops/deepin/library/qt5platform-plugins/default.nix:12]: na, then keep it like it is.

**Pull request merged**